src/WinRT.Runtime/GuidGenerator.cs

```diff
@@ -342,15 +346,37 @@ internal static Guid CreateIIDForGenericType(string signature)
                 return encode_guid(sha.ComputeHash(data));
             }
 #else
-            var maxBytes = UTF8Encoding.UTF8.GetMaxByteCount(signature.Length);
-
-            var data = new byte[16 /* Number of bytes in a GUID */ + maxBytes];
-            Span<byte> dataSpan = data;
-            wrt_pinterface_namespace.TryWriteBytes(dataSpan);
-            var numBytes = UTF8Encoding.UTF8.GetBytes(signature, dataSpan[16..]);
-            data = data[..(16 + numBytes)];
-
-            return encode_guid(SHA1.HashData(data));
+#nullable enable
+            // Get the maximum UTF8 byte size and allocate a buffer for the encoding.
+            // If the minimum buffer is small enough, we can stack-allocate it.
+            // The 512 threshold is the smallest one that will contain generally all
+            // enum types names. Eg. we'd get signature strings like this:
+            // 'pinterface({61c17706-2d65-11e0-9ae8-d48564015472};enum(Windows.UI.Xaml.Visibility;u4))'.
+            // Which would result in a number of bytes ~280, including the GUID size.
+            int maxUtf8ByteCount = Encoding.UTF8.GetMaxByteCount(signature.Length);
+            int minimumPooledLength = 16 /* Number of bytes in a GUID */ + maxUtf8ByteCount;
+            byte[]? utf8BytesFromPool = null;
+            Span<byte> utf8Bytes = minimumPooledLength <= 512
+                ? stackalloc byte[512]
+                : (utf8BytesFromPool = ArrayPool<byte>.Shared.Rent(minimumPooledLength));
+
+            wrt_pinterface_namespace.TryWriteBytes(utf8Bytes);
+
+            int encodedUtf8BytesWritten = Encoding.UTF8.GetBytes(signature, utf8Bytes[16..]);
+            Span<byte> sha1Bytes = stackalloc byte[SHA1.HashSizeInBytes];
+
+            // Hash the encoded signature (the bytes written are guaranteed to always match the span)
+            _ = SHA1.HashData(utf8Bytes[..(16 + encodedUtf8BytesWritten)], sha1Bytes);
+
+            Guid iid = encode_guid(sha1Bytes);
+
+            // Before exiting, make sure to return the array from the pool, if we rented one
+            if (utf8BytesFromPool is not null)

```

Comment:
I assume you didn't put this in a try finally because we assume HashData won't throw as we ensure correct size?

Reply:
Partly because of that (as in, I don't expect this to ever throw), but also because I'm following the security guidance from the .NET folks about not returning pooled arrays from `finally` blocks. The idea being that in case there's an exception, you shouldn't make assumptions on the state of the program (eg. perhaps you started some work somewhere that took a reference to the rented array and that wasn't correctly stopped, because of an exception, etc.), therefore in those cases you should just throw away the array instead, and only return them in normal circumstances (ie. if no exception is thrown).